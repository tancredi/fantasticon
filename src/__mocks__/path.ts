const _path = jest.requireActual('path');
const _relative = _path.relative;

const projectDir = _path.resolve(__dirname, '../../');

const resolve = (...paths: string[]) => {
  let path = '';

  for (const cur of paths) {
    path = !path ? normalize(cur) : _path.join(path, normalize(cur));
  }

  if (path.startsWith(projectDir)) {
    path = _path.join('/root/project', path.substr(projectDir.length));
  } else if (path.startsWith('/') && !path.startsWith('/root')) {
    path = _path.join('/root/', path.substr(1));
  } else if (path.startsWith('./')) {
    path = _path.join('/root/project/', path.slice(2));
  } else if (!path.startsWith('/')) {
    path = _path.join('/root/project/', path);
  }

  return normalize(path);
};

const relative = (a: string, b: string) =>
  normalize(_relative(_path.normalize(a), _path.normalize(b)));

const join = (...segments: string[]): string => {
  const trimmed: string[] = [];

  for (let [i, current] of segments.entries()) {
    const isFirst = i === 0;
    const isLast = i === segments.length - 1;

    if (!isFirst && current.startsWith('/')) {
      current = current.substr(1);
    }

    if (!isLast && current.endsWith('/')) {
      current = current.substring(0, current.length - 1);
    }

    trimmed.push(current);
  }

  return trimmed.join('/');
};

const normalize = (path: string) => path.replace(/\\/g, '/');

const isAbsolute = (path: string) => path.startsWith('/root');

module.exports = { resolve, relative, join, isAbsolute, normalize };
