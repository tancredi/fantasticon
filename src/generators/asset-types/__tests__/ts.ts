import { join, dirname } from 'path';
import tsGen from '../ts';

const mockAssets = {
  foo: {
    id: 'foo',
    absolutePath: '/path/there/foo.svg',
    relativePath: '../there/foo.svg'
  },
  bar: {
    id: 'bar',
    absolutePath: '/path/there/bar.svg',
    relativePath: '../there/bar.svg'
  }
};

const mockCodepoints = { foo: 4265, bar: 1231 };

const mockOptions = {
  name: 'my-icons-set',
  assets: mockAssets,
  codepoints: mockCodepoints,
  inputDir: join(dirname(__filename), 'fixtures'),
  outputDir: '/dev/null'
} as any;

const cleanWhiteSpace = (subject: string): string =>
  subject.replace(/\n+/, '').replace(/\s+/g, ' ');

const getCleanGen = async (options = {}) =>
  cleanWhiteSpace(
    (await tsGen.generate({ ...mockOptions, ...options }, null)) as string
  );

describe('`TS` asset generator', () => {
  test('renders expected TypeScript module content', async () => {
    expect(await tsGen.generate(mockOptions, null)).toMatchSnapshot();
  });

  test('correctly renders type declaration', async () => {
    expect(await getCleanGen()).toContain(
      'export type MyIconsSetId = | "foo" | "bar";'
    );
  });

  test('correctly enum declaration', async () => {
    expect(await getCleanGen()).toContain(
      'export enum MyIconsSet { Foo = "foo", Bar = "bar", }'
    );
  });

  test('correctly codepoints declaration', async () => {
    expect(await getCleanGen()).toContain(
      'export const MY_ICONS_SET_CODEPOINTS: { [key in MyIconsSet]: string }' +
        ' = { [MyIconsSet.Foo]: "4265", [MyIconsSet.Bar]: "1231", };'
    );
  });

  test('generates single quotes if format option passed', async () => {
    expect(
      await tsGen.generate(
        {
          ...mockOptions,
          formatOptions: { ts: { singleQuotes: true } }
        },
        null
      )
    ).toMatchSnapshot();
  });

  test('generates no key string literal type if option passed like that', async () => {
    const result = await tsGen.generate(
      {
        ...mockOptions,
        formatOptions: { ts: { types: ['constant', 'enum'] } }
      },
      null
    );
    const cleanResult = cleanWhiteSpace(result as string);
    expect(result).toMatchSnapshot();

    expect(cleanResult).not.toContain(
      'export type MyIconsSetKey = | "Foo" | "Bar";'
    );
    expect(cleanResult).not.toContain(
      'export type MyIconsSetId = | "foo" | "bar";'
    );
  });

  test('generates constant with literalId if no enum generated', async () => {
    const result = await tsGen.generate(
      {
        ...mockOptions,
        formatOptions: { ts: { types: ['constant', 'literalId'] } }
      },
      null
    );
    const cleanResult = cleanWhiteSpace(result as string);

    expect(result).toMatchSnapshot();
    expect(cleanResult).toContain(
      'export const MY_ICONS_SET_CODEPOINTS: { [key in MyIconsSetId]: string }'
    );

    expect(cleanResult).not.toContain(
      'export type MyIconsSetKey = | "Foo" | "Bar";'
    );
    expect(cleanResult).not.toContain('export enum MyIconsSet');
  });

  test('generates constant with literalKey if no enum generated', async () => {
    const result = await tsGen.generate(
      {
        ...mockOptions,
        formatOptions: { ts: { types: ['constant', 'literalKey'] } }
      },
      null
    );
    const cleanResult = cleanWhiteSpace(result as string);

    expect(result).toMatchSnapshot();
    expect(cleanResult).toContain(
      'export const MY_ICONS_SET_CODEPOINTS: { [key in MyIconsSetKey]: string }'
    );

    expect(cleanResult).not.toContain(
      'export type MyIconsSetId = | "foo" | "bar";'
    );
    expect(cleanResult).not.toContain('export enum MyIconsSet');
  });

  test('generates constant only', async () => {
    const result = await tsGen.generate(
      {
        ...mockOptions,
        formatOptions: { ts: { types: ['constant'] } }
      },
      null
    );
    const cleanResult = cleanWhiteSpace(result as string);

    expect(result).toMatchSnapshot();
    expect(cleanResult).toContain(
      'export const MY_ICONS_SET_CODEPOINTS: Record<string, string>'
    );

    expect(cleanResult).not.toContain(
      'export type MyIconsSetId = | "foo" | "bar";'
    );
    expect(cleanResult).not.toContain('export enum MyIconsSet');
  });

  test('prevents enum keys that start with digits', async () => {
    const result = await tsGen.generate(
      {
        ...mockOptions,
        assets: { 1234: mockAssets.foo, 5678: mockAssets.bar }
      },
      null
    );
    const cleanResult = cleanWhiteSpace(result as string);

    expect(result).toMatchSnapshot();
    expect(cleanResult).toContain(
      'export type MyIconsSetId = | "1234" | "5678";'
    );
    expect(cleanResult).toContain(
      'export type MyIconsSetKey = | "i1234" | "i5678";'
    );
    expect(cleanResult).toContain(
      'export enum MyIconsSet { i1234 = "1234", i5678 = "5678", }'
    );
  });

  test('prevents enum keys that start with digits when digits and chars', async () => {
    const result = await tsGen.generate(
      {
        ...mockOptions,
        assets: {
          '1234asdf': mockAssets.foo,
          '5678ab': mockAssets.bar,
          foo: mockAssets.foo
        }
      },
      null
    );
    const cleanResult = cleanWhiteSpace(result as string);

    expect(result).toMatchSnapshot();
    expect(cleanResult).toContain(
      'export type MyIconsSetId = | "1234asdf" | "5678ab" | "foo";'
    );
    expect(cleanResult).toContain(
      'export type MyIconsSetKey = | "i1234asdf" | "i5678ab" | "Foo";'
    );
    expect(cleanResult).toContain(
      'export enum MyIconsSet { i1234asdf = "1234asdf", i5678ab = "5678ab", Foo = "foo", }'
    );
  });
});
