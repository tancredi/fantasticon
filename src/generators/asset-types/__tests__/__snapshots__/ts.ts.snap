// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`TS\` asset generator generates constant only 1`] = `
"export const MY_ICONS_SET_CODEPOINTS: Record<string, string> = {
  "foo": "4265",
  "bar": "1231",
};
"
`;

exports[`\`TS\` asset generator generates constant with literalId if no enum generated 1`] = `
"export type MyIconsSetId =
  | "foo"
  | "bar";

export const MY_ICONS_SET_CODEPOINTS: { [key in MyIconsSetId]: string } = {
  "foo": "4265",
  "bar": "1231",
};
"
`;

exports[`\`TS\` asset generator generates constant with literalKey if no enum generated 1`] = `
"export type MyIconsSetKey =
  | "Foo"
  | "Bar";

export const MY_ICONS_SET_CODEPOINTS: { [key in MyIconsSetKey]: string } = {
  "foo": "4265",
  "bar": "1231",
};
"
`;

exports[`\`TS\` asset generator generates no key string literal type if option specifies it 1`] = `
"export enum MyIconsSet {
  Foo = "foo",
  Bar = "bar",
}

export const MY_ICONS_SET_CODEPOINTS: { [key in MyIconsSet]: string } = {
  [MyIconsSet.Foo]: "4265",
  [MyIconsSet.Bar]: "1231",
};
"
`;

exports[`\`TS\` asset generator generates single quotes if format option passed 1`] = `
"export type MyIconsSetId =
  | 'foo'
  | 'bar';

export type MyIconsSetKey =
  | 'Foo'
  | 'Bar';

export enum MyIconsSet {
  Foo = 'foo',
  Bar = 'bar',
}

export const MY_ICONS_SET_CODEPOINTS: { [key in MyIconsSet]: string } = {
  [MyIconsSet.Foo]: '4265',
  [MyIconsSet.Bar]: '1231',
};
"
`;

exports[`\`TS\` asset generator prevents enum keys that start with digits 1`] = `
"export type MyIconsSetId =
  | "1234"
  | "5678";

export type MyIconsSetKey =
  | "i1234"
  | "i5678";

export enum MyIconsSet {
  i1234 = "1234",
  i5678 = "5678",
}

export const MY_ICONS_SET_CODEPOINTS: { [key in MyIconsSet]: string } = {
  [MyIconsSet.i1234]: "undefined",
  [MyIconsSet.i5678]: "undefined",
};
"
`;

exports[`\`TS\` asset generator prevents enum keys that start with digits when digits and chars 1`] = `
"export type MyIconsSetId =
  | "1234asdf"
  | "5678ab"
  | "foo";

export type MyIconsSetKey =
  | "i1234asdf"
  | "i5678ab"
  | "Foo";

export enum MyIconsSet {
  i1234asdf = "1234asdf",
  i5678ab = "5678ab",
  Foo = "foo",
}

export const MY_ICONS_SET_CODEPOINTS: { [key in MyIconsSet]: string } = {
  [MyIconsSet.i1234asdf]: "undefined",
  [MyIconsSet.i5678ab]: "undefined",
  [MyIconsSet.Foo]: "4265",
};
"
`;

exports[`\`TS\` asset generator renders expected TypeScript module content 1`] = `
"export type MyIconsSetId =
  | "foo"
  | "bar";

export type MyIconsSetKey =
  | "Foo"
  | "Bar";

export enum MyIconsSet {
  Foo = "foo",
  Bar = "bar",
}

export const MY_ICONS_SET_CODEPOINTS: { [key in MyIconsSet]: string } = {
  [MyIconsSet.Foo]: "4265",
  [MyIconsSet.Bar]: "1231",
};
"
`;
